pirouette-plutusir/src/Language/Pirouette/PlutusIR/SMT.hs

```diff
@@ -0,0 +1,297 @@
+    | Just _ <- termIsMeta e = True
+  isStuckBuiltin (App (Free (Builtin op)) args)
+    | op `elem` plutusIRBasicOps || op `elem` plutusIRBasicRels,
+      all isArg args -- this ensures that we only have TermArg
+      =
+      let args' = map (\(TermArg a) -> a) args
+       in all isStuckBuiltin args' && not (all termIsConstant args')
+  isStuckBuiltin _ = False
+
+trPIRType :: PIRBuiltinType -> PureSMT.SExpr
+trPIRType PIRTypeInteger = PureSMT.tInt
+trPIRType PIRTypeString = PureSMT.tString
+trPIRType PIRTypeByteString = PureSMT.tString
+
+-- TODO Implement remaining constants
+trPIRConstant :: PIRConstant -> PureSMT.SExpr
+trPIRConstant (PIRConstInteger n) = PureSMT.int n
+trPIRConstant (PIRConstByteString bs) = PureSMT.text $ T.decodeUtf8 $ B64.encode bs
+trPIRConstant (PIRConstString txt) = PureSMT.text txt
+
+-- | These operations can be fully evaluated
+-- if the given arguments are constants.
+-- For example, @3 + 4@ is *not* stuck.
+plutusIRBasicOps :: [PIRDefaultFun]
+plutusIRBasicOps =
+  [ AddInteger,
+    SubtractInteger,
+    MultiplyInteger,
+    DivideInteger,
+    ModInteger,
+    QuotientInteger,
+    RemainderInteger,
+    AppendString,
+    AppendByteString,
+    ConsByteString,
+    IndexByteString,
+    SliceByteString
+  ]
+
+-- | These relations can be fully evaluated
+-- if the given arguments are constants.
+-- For example, @3 < 4@ is *not* stuck.
+plutusIRBasicRels :: [PIRDefaultFun]
+plutusIRBasicRels =
+  [ EqualsInteger,
+    LessThanInteger,
+    LessThanEqualsInteger,
+    EqualsByteString,
+    EqualsString,
+    LessThanByteString,
+    LessThanEqualsByteString
+  ]
+
+trPIRFun :: PIRDefaultFun -> [PureSMT.SExpr] -> Maybe PureSMT.SExpr
+
+-- TODO Implement remaining builtins: those used by the "Auction" example
+-- validator are marked with an [A] and as commented out lines in the
+-- code afterwards
+
+-- ** Hash and signatures **
+
+--
+--     Sha2_256
+--     Sha3_256
+--     Blake2b_256
+--     VerifySignature
+--
+
+-- ** Strings and encoding **
+
+--
+--     EncodeUtf8
+--     DecodeUtf8
+--
+
+-- Relations returning booleans actually need to instruct the solver
+-- to branch, their interpretation is handled in 'branchesBuiltinTerm'
+trPIRFun EqualsInteger _ = Nothing
+trPIRFun LessThanInteger _ = Nothing
+trPIRFun LessThanEqualsInteger _ = Nothing
+trPIRFun EqualsByteString _ = Nothing
+trPIRFun EqualsString _ = Nothing
+-- Unary
+trPIRFun op [x] =
+  case op of
+    Trace -> Just $ PureSMT.List [x]
+    _ ->
+      error $
+        "Translate builtin to SMT: "
+          <> show op
+          <> " is not an implemented unary operator/function"
+-- Binary operations and relations
+trPIRFun op [x, y] =
+  case op of
+    -- integer operations
+    AddInteger -> Just $ PureSMT.add x y
+    SubtractInteger -> Just $ PureSMT.sub x y
+    MultiplyInteger -> Just $ PureSMT.mul x y
+    -- divMod and quotRem work differently
+    -- when not exact, but this is a good approximation
```

Comment:
If the difference is in rounding towards -inf vs 0, we can express one of them using an explicit sign check, I guess. It'll look too ugly though, so not sure if that's worth the effort.